utils/wasm-gen/src/syscalls.rs

```diff
@@ -46,159 +32,62 @@ pub struct CallInfo {
 }
 
 impl CallInfo {
-    pub fn new(
-        config: &GearConfig,
-        signature: SysCallSignature,
-        frequency: Ratio,
-        skip_memory_array: bool,
-    ) -> Self {
+    pub fn new(call_signature: CallSignature, params_config: &SysCallsParamsConfig) -> Self {
+        let signature = call_signature.signature();
         Self {
             params: signature.params.iter().copied().map(Into::into).collect(),
             results: signature.results.to_vec(),
-            frequency,
-            parameter_rules: Self::into_parameter_rules(
-                config,
-                signature.params,
-                skip_memory_array,
+            parameter_rules: process_sys_call_params(
+                call_signature.name(),
+                &signature.params,
+                params_config,
             ),
         }
     }
 
     pub fn func_type(&self) -> FunctionType {
         FunctionType::new(self.params.clone(), self.results.clone())
     }
-
-    fn into_parameter_rules(
-        config: &GearConfig,
-        parameters: Vec<ParamType>,
-        skip_memory_array: bool,
-    ) -> Vec<Parameter> {
-        let mut rules = Vec::with_capacity(parameters.len());
-        for parameter in parameters.into_iter() {
-            match parameter {
-                ParamType::Size => match rules.last_mut() {
-                    None => rules.push((parameter, false)),
-                    Some((first, memory_array)) => match (first, *memory_array) {
-                        (ParamType::Ptr, false) if !skip_memory_array => *memory_array = true,
-                        _ => rules.push((parameter, false)),
-                    },
-                },
-
-                _ => rules.push((parameter, false)),
-            }
-        }
-
-        rules
-            .into_iter()
-            .map(|(arg_type, memory_array)| match memory_array {
-                true => Parameter::MemoryArray,
-                false => match arg_type {
-                    ParamType::Ptr => Parameter::MemoryValue,
-                    ParamType::Alloc => Parameter::Alloc,
-                    _ => Parameter::Value {
-                        value_type: arg_type.into(),
-                        rule: param_to_rule(arg_type, config),
-                    },
-                },
-            })
-            .collect()
-    }
 }
 
-/// Newtype for additional validation of calls arguments.
-///
-/// Parameters describing memory access should have correct values
-/// if required. For example, offset + length < memory_size for arrays.
-#[derive(Debug)]
-pub enum Parameter {
-    /// Some value with its type and generating rule.
-    Value {
-        value_type: ValueType,
-        rule: ParamRule,
-    },
-    /// Offset and length in memory for an array. Both have type i32.
-    MemoryArray,
-    /// Pointer in memory for some primitive type value.
-    MemoryValue,
-    /// Argument to `alloc` syscall.
-    Alloc,
+pub enum CallSignature {
```

Comment:
This is temporary and was introduced not to break current general design of the generator. With upcoming PRs will be removed.